File: book/petrol-pump/java/chapter4/section7/ShowDollarsPump.java
package chapter4.section7;

import pump.*;
import chapter4.section4.LifeCycle;
import nz.sodium.*;
import java.util.Optional;

public class ShowDollarsPump implements Pump {
    public Outputs create(Inputs inputs) {
        LifeCycle lc = new LifeCycle(inputs.sNozzle1,
                                     inputs.sNozzle2,
                                     inputs.sNozzle3);
        Fill fi = new Fill(lc.sStart.map(u -> Unit.UNIT),
                           inputs.sFuelPulses, inputs.calibration,
                           inputs.price1, inputs.price2, inputs.price3,
                           lc.sStart);
        return new Outputs()
            .setDelivery(lc.fillActive.map(
                of ->
                    of.equals(Optional.of(Fuel.ONE))   ? Delivery.FAST1 :
                    of.equals(Optional.of(Fuel.TWO))   ? Delivery.FAST2 :
                    of.equals(Optional.of(Fuel.THREE)) ? Delivery.FAST3 : 
                                                         Delivery.OFF))
            .setSaleCostLCD(fi.dollarsDelivered.map(
                q -> Formatters.formatSaleCost(q)))
            .setSaleQuantityLCD(fi.litersDelivered.map(
                q -> Formatters.formatSaleQuantity(q)))
            .setPriceLCD1(priceLCD(lc.fillActive, fi.price, Fuel.ONE,
                inputs))
            .setPriceLCD2(priceLCD(lc.fillActive, fi.price, Fuel.TWO,
                inputs))
            .setPriceLCD3(priceLCD(lc.fillActive, fi.price, Fuel.THREE,
                inputs));
    }

    public static Cell<String> priceLCD(
            Cell<Optional<Fuel>> fillActive,
            Cell<Double> fillPrice,
            Fuel fuel,
            Inputs inputs) {
        Cell<Double> idlePrice;
        switch (fuel) {
            case ONE:   idlePrice = inputs.price1; break;
            case TWO:   idlePrice = inputs.price2; break;
            case THREE: idlePrice = inputs.price3; break;
            default:    idlePrice = null;
        }
        return fillActive.lift(fillPrice, idlePrice,
        	(oFuelSelected, fillPrice_, idlePrice_) ->
				oFuelSelected.isPresent()
					? oFuelSelected.get() == fuel
										  ? Formatters.formatPrice(fillPrice_)
										  : ""
					: Formatters.formatPrice(idlePrice_));
    }
}

File: book/petrol-pump/c#/PetrolPump/Chapter4/Section7/ShowDollarsPump.cs
using PetrolPump.Chapter4.Section4;
using Sodium.Frp;
using Sodium.Functional;

namespace PetrolPump.Chapter4.Section7
{
    public class ShowDollarsPump : IPump
    {
        public Outputs Create(Inputs inputs)
        {
            LifeCycle lc = new LifeCycle(inputs.SNozzle1,
                inputs.SNozzle2,
                inputs.SNozzle3);
            Fill fi = new Fill(lc.SStart.Map(_ => Unit.Value),
                inputs.SFuelPulses, inputs.Calibration,
                inputs.Price1, inputs.Price2, inputs.Price3,
                lc.SStart);
            return new Outputs()
                .SetDelivery(lc.FillActive.Map(
                    m =>
                        m.Equals(Maybe.Some(Fuel.One)) ? Delivery.Fast1 :
                            m.Equals(Maybe.Some(Fuel.Two)) ? Delivery.Fast2 :
                                m.Equals(Maybe.Some(Fuel.Three)) ? Delivery.Fast3 :
                                    Delivery.Off))
                .SetSaleCostLcd(fi.DollarsDelivered.Map(Formatters.FormatSaleCost))
                .SetSaleQuantityLcd(fi.LitersDelivered.Map(Formatters.FormatSaleQuantity))
                .SetPriceLcd1(PriceLcd(lc.FillActive, fi.Price, Fuel.One, inputs))
                .SetPriceLcd2(PriceLcd(lc.FillActive, fi.Price, Fuel.Two, inputs))
                .SetPriceLcd3(PriceLcd(lc.FillActive, fi.Price, Fuel.Three, inputs));
        }

        public static Cell<string> PriceLcd(
            Cell<Maybe<Fuel>> fillActive,
            Cell<double> fillPrice,
            Fuel fuel,
            Inputs inputs)
        {
            Cell<double> idlePrice;
            switch (fuel)
            {
                case Fuel.One:
                    idlePrice = inputs.Price1;
                    break;
                case Fuel.Two:
                    idlePrice = inputs.Price2;
                    break;
                case Fuel.Three:
                    idlePrice = inputs.Price3;
                    break;
                default:
                    idlePrice = null;
                    break;
            }
            return fillActive.Lift(fillPrice, idlePrice,
                (fuelSelectedMaybe, fillPriceLocal, idlePriceLocal) =>
                    fuelSelectedMaybe.Match(f => f == fuel ? Formatters.FormatPrice(fillPriceLocal, 4) : string.Empty, () => Formatters.FormatPrice(idlePriceLocal, 4)));
        }
    }
}